### ISSUE TITLE
could not run executable : Permission denied (os error 13)

### ISSUE THREAD

**Issue by nickdesaulniers** [KyleMayes/clang-sys#138]
Hello,
I'm trying to review patches to the Linux kernel for adding Rust support to the build system. It seems that they are using bindgen, which I think depends on this crate.

It looks like the build system is invoking:
```
bindgen rust/helpers.c --blacklist-type '.*' --whitelist-var '' --whitelist-function 'rust_helper_.*' --use-core --with-derive-default --ctypes-prefix c_types --no-debug '.*' --size_t-is-usize -o rust/bindings_helpers_generated.rs -- -Wp,-MMD,rust/.bindings_helpers_generated.rs.d -nostdinc -I./arch/x86/include -I./arch/x86/include/generated -I./include -I./arch/x86/include/uapi -I./arch/x86/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/compiler-version.h -include ./include/linux/kconfig.h -include ./include/linux/compiler_types.h -D__KERNEL__ -Qunused-arguments -fmacro-prefix-map=./= -Wall -Wundef -Werror=strict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -fshort-wchar -fno-PIE -Werror=implicit-function-declaration -Werror=implicit-int -Werror=return-type -Wno-format-security -std=gnu89 --target=x86_64-linux-gnu -fintegrated-as -Werror=unknown-warning-option -Werror=ignored-optimization-argument -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx -fcf-protection=none -m64 -falign-loops=1 -mno-80387 -mno-fp-ret-in-387 -mstack-alignment=8 -mskip-rax-setup -mtune=generic -mno-red-zone -mcmodel=kernel -Wno-sign-compare -fno-asynchronous-unwind-tables -mretpoline-external-thunk -fno-delete-null-pointer-checks -Wno-frame-address -Wno-address-of-packed-member -O2 -Wframe-larger-than=2048 -fstack-protector-strong -Wimplicit-fallthrough -Wno-gnu -mno-global-merge -Wno-unused-but-set-variable -Wno-unused-const-variable -fomit-frame-pointer -ftrivial-auto-var-init=zero -enable-trivial-auto-var-init-zero-knowing-it-will-be-removed-from-clang -fno-stack-clash-protection -Wdeclaration-after-statement -Wvla -Wno-pointer-sign -Wno-array-bounds -fno-strict-overflow -fno-stack-check -Werror=date-time -Werror=incompatible-pointer-types -Wno-initializer-overrides -Wno-format -Wno-sign-compare -Wno-format-zero-length -Wno-pointer-to-enum-cast -Wno-tautological-constant-out-of-range-compare -DKBUILD_MODFILE='"rust/bindings_helpers_generated"' -DKBUILD_BASENAME='"bindings_helpers_generated"' -DKBUILD_MODNAME='"bindings_helpers_generated"' -D__KBUILD_MODNAME=kmod_bindings_helpers_generated -I./rust/ -DMODULE; sed -Ei 's/pub fn rust_helper_([a-zA-Z0-9_]*)/#[link_name="rust_helper_\1"]\n    pub fn \1/g' rust/bindings_helpers_generated.rs
```
which is producing the following error for me:
```        
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "could not run executable `/android0/llvm-project/clang`: Permission denied (os error 13)"', /usr/local/google/home/ndesaulniers/.cargo/registry/src/github.com-1ecc6299db9ec823/clang-sys-1.0.3/src/support.rs:165:58
```

This may be a special case for me. I am an LLVM developer, and thus have a copy of llvm-project checked out AND the resulting clang binary I've built locally in my `$PATH`.  It looks like clang-sys is finding my source checkout in `/android0/llvm-project/`, but my clang binary is actually located at:
```
$ which clang
/android0/llvm-project/llvm/build/bin/clang
```
hence the failure to execute `/android0/llvm-project/clang`.

Also, top of tree LLVM is clang-14 FWIW; I saw something in the readme about clang-13 support. Dunno if that's also an issue yet.

cc @ojeda @wedsonaf

**Comment by KyleMayes** [comment]
So the file that could not be executed (`/android0/llvm-project/clang`)  is a directory?

**Comment by nickdesaulniers** [comment]
Yes. It's the source for `clang` in the `llvm-project` [repository](https://github.com/llvm/llvm-project/tree/main/clang).
```
$ file /android0/llvm-project/clang
/android0/llvm-project/clang: directory
```

**Comment by Xuanwo** [comment]
https://github.com/KyleMayes/clang-sys/blob/8090d6853624a04b859e8ab2e467bae88a8d4ef6/src/support.rs#L142-L149

Is this issue still active? It seems clang-sys has improved the executable check logic a lot since v1.0.3.

**Comment by nickdesaulniers** [comment]
The [v7 Linux kernel patches](https://lore.kernel.org/lkml/20220523020209.11810-1-ojeda@kernel.org/) were just sent out, and I still cannot test them locally.  Looks like they force the use of bindgen v0.56.0.

```
RUST_BACKTRACE=1 bindgen scripts/rust-is-available-bindgen-libclang.h
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "could not run executable `/android0/llvm-project/clang`: Permission denied (os error 13)"', /usr/local/google/home/ndesaulniers/.cargo/registry/src/github.com-1ecc6299db9ec823/clang-sys-1.0.3/src/support.rs:165:58
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::result::unwrap_failed
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1616:5
   3: clang_sys::support::run_clang
   4: clang_sys::support::parse_version
   5: clang_sys::support::Clang::find
   6: bindgen::Builder::generate
   7: std::panicking::try
   8: bindgen::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
It's still trying to execute my source directory for LLVM.  If I forcibly upgrade to [0.59.2](https://crates.io/crates/bindgen/0.59.2), the issue persists. (Looks like newer bindgen uses older revision of clang-sys?)
```
RUST_BACKTRACE=1 bindgen scripts/rust-is-available-bindgen-libclang.h
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "could not run executable `/android0/llvm-project/clang`: Permission denied (os error 13)"', /usr/local/google/home/ndesaulniers/.cargo/registry/src/github.com-1ecc6299db9ec823/clang-sys-1.2.0/src/support.rs:202:58
stack backtrace:
   0: rust_begin_unwind
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/panicking.rs:143:14
   2: core::result::unwrap_failed
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/result.rs:1749:5
   3: clang_sys::support::run_clang
   4: clang_sys::support::parse_version
   5: clang_sys::support::Clang::find
   6: bindgen::Builder::generate
   7: std::panicking::try
   8: bindgen::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

https://github.com/rust-lang/rust-bindgen/commit/704b9eece3b98ca40af109d22ecd8afa55af2c62

I'm not sure how to forcibly upgrade clang-sys and have my install of bindgen use that.

**Comment by nickdesaulniers** [comment]
idk if I need to file a request for bindgen to upgrade the version of clang-sys they're using?
https://github.com/rust-lang/rust-bindgen/blob/5a01c551993e56d20240ef64d0ec78cd4195855d/Cargo.toml#L53
just says version 1. IDK if cargo uses semver, but I don't understand why my install of bindgen is using such an old version of clang-sys (1.0.3 IIUC from the trace provided above).

**Comment by Xuanwo** [comment]
> idk if I need to file a request for bindgen to upgrade the version of clang-sys they're using?

No, I don't think so.

`bindgen` locked the version of `clang-sys` to: https://github.com/rust-lang/rust-bindgen/blob/5a01c551993e56d20240ef64d0ec78cd4195855d/Cargo.lock#L75-L84

```toml
[[package]]
name = "clang-sys"
version = "1.2.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "853eda514c284c2287f4bf20ae614f8781f40a81d32ecda6e91449304dfe077c"
dependencies = [
 "glob",
 "libc",
 "libloading",
]
```

Which has already fixed this problem (since 1.0.3)

**Comment by Xuanwo** [comment]
> Yes. It's the source for `clang` in the `llvm-project` [repository](https://github.com/llvm/llvm-project/tree/main/clang).
> 
> ```
> $ file /android0/llvm-project/clang
> /android0/llvm-project/clang: directory
> ```

Hi, can you share the expected clang path? Thanks.

**Comment by ojeda** [comment]
> Hi, can you share the expected clang path? Thanks.

I have no access to Nick's environment, but if I understand correctly the first message here, his real binary is at `/android0/llvm-project/llvm/build/bin/clang`.

I suggested trying to point to the real paths with the environment variables: https://lore.kernel.org/rust-for-linux/CANiq72=dLnCpCLOKJ6+x+D0AV=tFNyy4hUEjd6KCOCw+eWbPkA@mail.gmail.com/

**Comment by nickdesaulniers** [comment]
bindgen 0.59.2 is using clang-sys 1.2.0.  The error I observe is from:
https://github.com/KyleMayes/clang-sys/blob/v1.2.0/src/support.rs#L202

> Which has already fixed this problem (since 1.0.3)

No, it has not. I'm using 1.2.0 and observing the error.

Clang::new still calls parse_version which calls run_clang without an is_executable check.  I don't think the error handling of run_clang is actually working as you expect.

**Comment by KyleMayes** [comment]
Fix released in v1.3.3.
Thanks!
